test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,394 @@
+    tail --pid="$pid" -f /dev/null
+}
+
+arecord_in_background () {
+    dlogi Capturing file "$1"
+    arecord -D "$CAP_HW" -f $CAP_FORMAT -r $CAP_RATE -c "$2" -d "$3" "$1"  2> /dev/null &
+    pid=$!
+    return $pid
+}
+
+aplay_in_background () {
+    if ! [ -f "$1" ]; then
+	dloge "Error: File $1 does not exist"
+	exit 1
+    fi
+
+    dlogi Playing file "$1"
+    aplay -D "$PLAY_HW" "$1"  2> /dev/null &
+    pid=$!
+    return $pid
+}
+
+do_preparations () {
+
+    # Check that TPLG is set
+    if [ -z "$TPLG" ]; then
+	dloge "Error: Need to set environment variable TPLG to the used SOF topology"
+	exit 1
+    fi
+
+    if [[ "$TPLG" != *"nocodec"* ]]; then
+	dlogi "Note: this test is executed only with nocodec topologies."
+	exit 0
+    fi
+
+    # Get max. channels count to use from capture device
+    # remove xargs after the trailing blank from tplgreader is fixed
+    CAP_CHANNELS=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d ch_max -v | xargs)
+    CAP_HW=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d dev -v | xargs)
+    PLAY_HW=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:playback" -d dev -v | xargs)
+    CAP_PGA=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d pga -v | xargs)
+    PLAY_PGA=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:playback" -d pga -v | xargs)
+    export CAP_CHANNELS
+    export CAP_HW
+    export PLAY_HW
+
+    dlogi Test uses "$CAP_CHANNELS" channels
+    dlogi Playback device is "$PLAY_HW"
+    dlogi Playback PGA is "$PLAY_PGA"
+    dlogi Capture device is "$CAP_HW"
+    dlogi Capture PGA is "$CAP_PGA"
+
+    # Find amixer controls for capture, error if more than one PGA
+    numpga=${#CAP_PGA[@]}
+    if [ "$numpga" != 1 ]; then
+	dloge Error: more than one capture PGA found.
+	exit 1
+    fi
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Volume)
+    search="name="
+    CAP_VOLUME=${tmp#*$search}
+    export CAP_VOLUME
+
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Switch)
+    search="name="
+    CAP_SWITCH=${tmp#*$search}
+    export CAP_SWITCH
+
+    dlogi Capture Volume name is "$CAP_VOLUME"
+    dlogi Capture Volume name is "$CAP_SWITCH"
+
+    # Check needed controls
+    amixer cget name="$CAP_VOLUME"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: failed capture volume get command"
+	exit $ret
+    fi
+
+    amixer cget name="$CAP_SWITCH"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: failed capture switch get command"
+	exit $ret
+    fi
+
+    for pga in $PLAY_PGA
+    do
+	tmp=$(amixer controls | grep "$pga" | grep Volume)
+	search="name="
+	play_volume=${tmp#*$search}
+	dlogi Set "$play_volume" to 100%
+	amixer cset name="$play_volume" 100%
+	ret=$?
+	if [ $ret -ne 0 ]; then
+	    dloge "Error: failed play volume set command"
+	    exit $ret
+	fi
+    done
+}
+
+generate_sine () {
+
+    dlogi Creating sine wave file
+    octave --silent --no-gui --eval "check_volume_levels('generate', '$APLAY_WAV')"
+    ret=$?
+    if [ $ret -ne 0 ]; then
```

Comment:
For compatibility with `set -e` (which is missing now)
```shell
octave --silent --no-gui --eval "check_volume_levels('generate', '$APLAY_WAV')" ||
  die "Error: failed sine wave generate"

```
Not sure how this is not triggering https://github.com/koalaman/shellcheck/wiki/SC2181 ?!

Reply:
Done.